HTML `pattern` attribute should set `v` flag for regular expressions

annevk opened the pull request:
#### 092303e2abb1e89254f64b301953ac0fbf36c681
<pre>
HTML `pattern` attribute should set `v` flag for regular expressions
<a href="https://bugs.webkit.org/show_bug.cgi?id=251675">https://bugs.webkit.org/show_bug.cgi?id=251675</a>
rdar://105268069

Reviewed by Yusuke Suzuki.

Make &lt;input pattern&gt; use Flags::UnicodeSets instead of Flags::Unicode
and encourage future callers of RegularExpression to do the same.

* LayoutTests/fast/forms/ValidityState-patternMismatch-expected.txt:
* LayoutTests/fast/forms/ValidityState-patternMismatch.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/constraints/form-validation-validity-patternMismatch-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/constraints/form-validation-validity-patternMismatch.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/constraints/reportValidity-crash.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/constraints/w3c-import.log:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/pattern_attribute-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/pattern_attribute.html:
* Source/JavaScriptCore/yarr/RegularExpression.cpp:
(JSC::Yarr::RegularExpression::Private::compile):
* Source/WebCore/html/BaseTextInputType.cpp:
(WebCore::BaseTextInputType::patternMismatch const):

Canonical link: <a href="https://commits.webkit.org/264245@main">https://commits.webkit.org/264245@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0bef94388e495d2ad7ee434fa358d4dcaab9b448

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7142 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7390 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7568 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8759 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7363 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7152 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8659 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7322 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10265 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7269 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7931 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6579 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8865 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/5285 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6498 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14241 "1 flakes 113 failures") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/6051 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6945 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6597 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/9474 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/6726 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7078 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/5787 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/7280 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6438 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/6407 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1687 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1691 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10637 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/7482 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6821 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1827 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

annevk commented on LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/pattern_attribute-expected.txt:
```diff
@@ -1,5 +1,11 @@
+PASS <input pattern> supports Unicode property escape syntax for properties of strings
+PASS <input pattern> supports set difference syntax
+PASS <input pattern> supports string literal syntax
+FAIL <input pattern> enables the RegExp v flag assert_true: <input pattern="[&amp;&amp;]" value="foo"> should be valid expected true got false
```
@mathiasbynens is this an existing WebKit bug with our v implementation or is something else amiss here?

mathiasbynens replied:
`/[&&]/v` is meant to throw an early `SyntaxError`, since `&&` matches [the `ClassSetReservedDoublePunctuator` production](https://arai-a.github.io/ecma262-compare/?pr=2418#ClassSetReservedDoublePunctuator) in this case. This is also given as an example in the proposal README: https://github.com/tc39/proposal-regexp-v-flag#how-is-the-v-flag-different-from-the-u-flag And since the pattern doesn’t compile, any input value is expected to be valid.

@msaboff, is this an implementation bug, or did you find an error in our proposed spec text?

annevk replied:
`new RegExp(/[&&]/v)` doesn't throw in WebKit so that does indeed appear to be a deeper issue. I think that also means that this is in principle ready for review, though we might not want to land this until we have interoperable v semantics.

msaboff replied:
This is an existing bug in the WebKit implementation that I happen to be working on.

mathiasbynens replied:
Thanks for confirming. I’ve written some more Test262 tests explicitly capturing these cases now: https://github.com/tc39/test262/pull/3820

Constellation reviewed (approved):
r=me

annevk committed:
HTML `pattern` attribute should set `v` flag for regular expressions
https://bugs.webkit.org/show_bug.cgi?id=251675
rdar://105268069

Reviewed by Yusuke Suzuki.

Make <input pattern> use Flags::UnicodeSets instead of Flags::Unicode
and encourage future callers of RegularExpression to do the same.

* LayoutTests/fast/forms/ValidityState-patternMismatch-expected.txt:
* LayoutTests/fast/forms/ValidityState-patternMismatch.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/constraints/form-validation-validity-patternMismatch-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/constraints/form-validation-validity-patternMismatch.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/constraints/reportValidity-crash.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/constraints/w3c-import.log:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/pattern_attribute-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/pattern_attribute.html:
* Source/JavaScriptCore/yarr/RegularExpression.cpp:
(JSC::Yarr::RegularExpression::Private::compile):
* Source/WebCore/html/BaseTextInputType.cpp:
(WebCore::BaseTextInputType::patternMismatch const):

Canonical link: https://commits.webkit.org/264245@main
